### PR TITLE
feature(scheduler): implement ClusterEventWithHint to filter out useless events

### PR DESCRIPTION
**Pull request by sanposhiho**
#### What type of PR is this?

/kind feature

#### What this PR does / why we need it:

The `EventsToRegister` in `EnqueueExtension` changed the return value from `ClusterEvent` to `ClusterEventWithHint`. `ClusterEventWithHint` allows each plugin to filter out more useless events via the callback function named `QueueingHintFn`.
When the scheduling queue receives a cluster event, before moving each Pod from unschedulable pod pool to activeQ/backoffQ, it will call QueueingHintFn of plugins that rejected each Pod in the previous scheduling cycle.
Depending on the value returned from QueueingHintFn, the scheduling queue changes how it queues each Pod:
- if more than one QueueingHintFn returns QueueImmediately, it queues Pod to activeQ.
- If no QueueingHintFn returns QueueImmediately and more than one plugin returns QueueAfterBackoff, it queues Pod to backoffQ if Pod is backing off, or to activeQ if Pod's backoff has already finished.
- If all QueueingHintFn return QueueSkip, it puts this pod back to the unschedulable pod pool

#### Which issue(s) this PR fixes:

Part of https://github.com/kubernetes/kubernetes/issues/114297 (PoC: https://github.com/kubernetes/kubernetes/pull/117844)

#### Special notes for your reviewer:

/hold

This is a big change, we should involve another approver to review.

#### Does this PR introduce a user-facing change?

```release-note
Action required for the custom scheduler plugin developers. 
Here's the breaking change in `EnqueueExtension` in the scheduling framework. 
The `EventsToRegister` in `EnqueueExtension` changed the return value from `ClusterEvent` to `ClusterEventWithHint`. `ClusterEventWithHint` allows each plugin to filter out more useless events via the callback function named `QueueingHintFn`.
When the scheduling queue receives a cluster event, before moving each Pod from unschedulable pod pool to activeQ/backoffQ, it will call QueueingHintFn of plugins that rejected each Pod in the previous scheduling cycle.
Depending on the value returned from QueueingHintFn, the scheduling queue changes how it queues each Pod:
- if more than one QueueingHintFn returns QueueImmediately, it queues Pod to activeQ.
- If no QueueingHintFn returns QueueImmediately and more than one plugin returns QueueAfterBackoff, it queues Pod to backoffQ if Pod is backing off, or to activeQ if Pod's backoff has already finished.
- If all QueueingHintFn return QueueSkip, it puts this pod back to the unschedulable pod pool

Having appropriate QueueingHintFn contributes to reducing useless retries and thus improves the overall scheduler's performance.

**How can I migrate?**

For backward compatibility, nil `QueueingHintFn` is treated as always returning QueueAfterBackoff. 
So, if you want to just keep the existing behavior, you can register `ClusterEventWithHint` with no `QueueingHintFn` in it. 
But, registering appropriate `QueueingHintFn` is, of course, better from a scheduling performance perspective.
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

```docs

```

### CODE REVIEW

**Comment by sanposhiho** [comment]
/triage accepted
/priority important-longterm

We want to prioritize this because it'll contribute to efficient enqueueing overall along with other changes for https://github.com/kubernetes/kubernetes/issues/114297.

File: pkg/scheduler/internal/queue/scheduling_queue.go [hunk 1]
```diff
@@ -336,6 +347,56 @@ func (p *PriorityQueue) Run(logger klog.Logger) {
...
 
+// isPodWorthRequeuing calls QueueingHintFn of only plugins registered in pInfo.unschedulablePlugins.
+// If any QueueingHintFn returns QueueImmediately, the scheduling queue is supposed to enqueue this Pod to activeQ.
+// If no QueueingHintFn returns QueueImmediately, but some return QueueAfterBackoff,
+// the scheduling queue is supposed to enqueue this Pod to activeQ/backoffQ depending on the remaining backoff time of the Pod.
+// If all QueueingHintFn returns QueueSkip, the scheduling queue enqueues the Pod back to unschedulable Pod pool
+// because no plugin changes the scheduling result via the event.
+func (p *PriorityQueue) isPodWorthRequeuing(logger klog.Logger, pInfo *framework.QueuedPodInfo, event framework.ClusterEvent, oldObj, newObj interface{}) framework.QueueingHint {
+	if pInfo.UnschedulablePlugins.Len() == 0 || event.IsWildCard() {
+		logger.V(6).Info("Worth requeuing because no unschedulable plugins", "pod", klog.KObj(pInfo.Pod))
```
Comment (alculquicondor) [inline pkg/scheduler/internal/queue/scheduling_queue.go:358]: was this existing logic?

Also, maybe worth different messages for empty `UnschedulablePlugins` and the wildcard.

File: pkg/scheduler/internal/queue/scheduling_queue.go [hunk 2]
```diff
@@ -336,6 +347,56 @@ func (p *PriorityQueue) Run(logger klog.Logger) {
...
+		return framework.QueueAfterBackoff
+	}
+
+	hintMap, ok := p.queueingHintMap[pInfo.Pod.Spec.SchedulerName]
+	if !ok {
+		logger.Error(nil, "No QueueingHintMap is registered for this profile", "profile", pInfo.Pod.Spec.SchedulerName, "pod", klog.KObj(pInfo.Pod))
```
Comment (alculquicondor) [inline pkg/scheduler/internal/queue/scheduling_queue.go:364]: uhm... is this possible? Would the pod have entered the scheduling queue at all?
If not, worth clarifying in a comment that this shouldn't be reachable in production.

File: pkg/scheduler/internal/queue/scheduling_queue.go [hunk 3]
```diff
@@ -336,6 +347,56 @@ func (p *PriorityQueue) Run(logger klog.Logger) {
...
+		return framework.QueueAfterBackoff
+	}
+
+	pod := pInfo.Pod
+	queueHint := framework.QueueSkip
+	for eve, hintfns := range hintMap {
```
Comment (alculquicondor) [inline pkg/scheduler/internal/queue/scheduling_queue.go:370]:
```suggestion
	for eventToMatch, hintfns := range hintMap {
```

File: pkg/scheduler/internal/queue/scheduling_queue.go (hunk 1 above)
Comment (sanposhiho) [reply pkg/scheduler/internal/queue/scheduling_queue.go:358]: Yes. In both cases, we always enqueue Pod to activeQ/backoffQ now.

- When `pInfo.UnschedulablePlugins.Len() == 0`
  - https://github.com/kubernetes/kubernetes/blob/336e46101c4451f504c884d65bb7b7b5f6f7f4d8/pkg/scheduler/internal/queue/scheduling_queue.go#L778
- When `event.IsWildCard()`
  - `podMatchesEvent` always returns `true` https://github.com/kubernetes/kubernetes/blob/336e46101c4451f504c884d65bb7b7b5f6f7f4d8/pkg/scheduler/internal/queue/scheduling_queue.go#L1151-L1153
  - and then https://github.com/kubernetes/kubernetes/blob/336e46101c4451f504c884d65bb7b7b5f6f7f4d8/pkg/scheduler/internal/queue/scheduling_queue.go#L778


> Also, maybe worth different messages for empty UnschedulablePlugins and the wildcard.

👍

File: pkg/scheduler/internal/queue/scheduling_queue.go (hunk 2 above)
Comment (sanposhiho) [reply pkg/scheduler/internal/queue/scheduling_queue.go:364]: > If not, worth clarifying in a comment that this shouldn't be reachable in production.

👌

File: pkg/scheduler/framework/types.go [hunk 4]
```diff
@@ -78,6 +78,57 @@ const (
 	WildCard              GVK = "*"
 )
 
+type ClusterEventWithHint struct {
+	Event ClusterEvent
+	// QueueingHintFn is executed for the plugin rejected by this plugin when the above Event happens,
+	// and filters out events to reduce useless retry of Pod's scheduling.
+	// It's the optional field. If you don't have any in this field,
+	// the scheduling of Pods will be always retried when this Event happens.
```
[the remark this answers is not in the document]
Comment (sanposhiho) [reply pkg/scheduler/framework/types.go:86]: Yes, correct. OK, will add that to the comment.

File: pkg/scheduler/framework/types.go [hunk 5]
```diff
@@ -78,6 +78,57 @@ const (
...
+
+	// QueueAfterBackoff implies that the Pod may be schedulable by the event,
+	// and worth retring the scheduling again after backoff.
+	QueueAfterBackoff
+
+	// QueueImmediately is returned only when it is highly possible that the Pod gets scheduled in the next scheduling.
+	// Note that you should return QueueImmediately when the plugin is 100% sure that the Pod gets scheduled in the next scheduling.
+	// Otherwise, it's detrimental to scheduling throughput.
+	// For example, when it depends on other plugin's decisions whether the Pod is schedulable or not by the event,
+	// the plugin should return QueueAfterBackoff instead of QueueImmediately.
```
[the remark this answers is not in the document]
Comment (sanposhiho) [reply pkg/scheduler/framework/types.go]: Sure, added two examples to explain when to return each.

File: pkg/scheduler/scheduler.go [hunk 6]
```diff
@@ -354,6 +356,33 @@ func New(ctx context.Context,
 	return sched, nil
```
[the remark this answers is not in the document]
Comment (sanposhiho) [reply pkg/scheduler/scheduler.go:354]: Good idea, changed as your advice

File: pkg/scheduler/scheduler_test.go [hunk 7]
```diff
@@ -643,3 +643,372 @@ func indexByPodAnnotationNodeName(obj interface{}) ([]string, error) {
...
+						continue
+					}
+				}
+			}
+		})
+	}
+}
+
+// Test_UnionedGVKs tests UnionedGVKs worked with buildQueueingHintMap.
+func Test_UnionedGVKs(t *testing.T) {
```
Comment (sanposhiho) [inline pkg/scheduler/scheduler_test.go:801]: This test proves currently even if we register any plugins, we register the same in event handlers. (It's existing behavior, not due to the change of this PR.)

It's because the default queue sort plugin etc don't implement EventsToRegister, but we apply default EventsToRegister on such plugins in NewFramework().
Probably, we don't need to do that when it comes to queuesort plugin or other re-enqueue unrelated plugins if they don't have EventsToRegister.
I'll follow up in another PR after this PR.

**Comment by sanposhiho** [comment]
@alculquicondor Thanks for the detailed review, I addressed them. 🙏

File: pkg/scheduler/internal/queue/scheduling_queue_test.go [hunk 8]
```diff
@@ -2315,3 +2253,269 @@ func mustNewPodInfo(pod *v1.Pod) *framework.PodInfo {
...
+				t.Errorf("isPodWorthRequeuing() = %v, want %v", actual, test.expected)
+			}
+			if count != test.expectedExecutionCount {
+				t.Errorf("isPodWorthRequeuing() executed queueing hint functions %v times, expected: %v", count, test.expectedExecutionCount)
+			}
+		})
+	}
+}
+
+func Test_requeuePodViaQueueingHint(t *testing.T) {
```
Comment (alculquicondor) [inline pkg/scheduler/internal/queue/scheduling_queue_test.go:2474]: Can you test `MoveAllToActiveOrBackoffQueue` instead? Preferably we shouldn't test non exported functions, as they could be refactored quite frequently.

We can make exceptions to this rule when a function is too complex, `isPodWorthRequeuing` falls into that.

File: pkg/scheduler/scheduler.go [hunk 9]
```diff
@@ -349,11 +349,36 @@ func New(ctx context.Context,
...
 }
 
+// defaultQueueingHintFn is the default queueing hint function.
+// It always returns QueueAfterBackoff as the queueing hint.
+var defaultQueueingHintFn = func(_ *v1.Pod, _, _ interface{}) framework.QueueingHint {
+	return framework.QueueAfterBackoff
+}
+
+func buildQueueingHintMap(es []framework.EnqueueExtensions) map[framework.ClusterEvent][]*internalqueue.QueueingHintFunction {
```
Comment (alculquicondor) [inline pkg/scheduler/scheduler.go:363]: Name this returning type.

File: pkg/scheduler/scheduler_test.go [hunk 10]
```diff
@@ -643,3 +643,376 @@ func indexByPodAnnotationNodeName(obj interface{}) ([]string, error) {
...
+	tests := []struct {
+		name    string
+		plugins []framework.Plugin
```
Comment (alculquicondor) [inline pkg/scheduler/scheduler_test.go:804]: could you add a test with the default profile? Only if it doesn't add a circular dependency.

**Review by alculquicondor — COMMENTED** [review]
Do we have sufficient coverage in unit or integration tests to give us confidence that we are not introducing regressions?
Could you produce a list of the tests that you believe give us this confidence?

**Comment by sanposhiho** [comment]
@alculquicondor 

> Could you produce a list of the tests that you believe give us this confidence?

unit tests:
- TestPriorityQueue_MoveAllToActiveOrBackoffQueue
  - https://github.com/sanposhiho/kubernetes/blob/6dd7a0b66be0f2d9d015134561f658f0d69e8c4d/pkg/scheduler/internal/queue/scheduling_queue_test.go#L764
  - covering the existing behavior of `MoveAllToActiveOrBackoffQueue`. (around enqueue via comparison of failure plugins and events)
- TestMoveAllToActiveOrBackoffQueue_PreEnqueueChecks
  - https://github.com/sanposhiho/kubernetes/blob/6dd7a0b66be0f2d9d015134561f658f0d69e8c4d/pkg/scheduler/internal/queue/scheduling_queue_test.go#L2180
  - covering the existing behavior of `MoveAllToActiveOrBackoffQueue` with preCheck.

integration tests:
- TestCoreResourceEnqueue
  - https://github.com/sanposhiho/kubernetes/blob/6dd7a0b66be0f2d9d015134561f658f0d69e8c4d/test/integration/scheduler/queue_test.go#L178
  - covering the existing behavior of re-queueing with core resources.
- TestCustomResourceEnqueue
  - https://github.com/sanposhiho/kubernetes/blob/6dd7a0b66be0f2d9d015134561f658f0d69e8c4d/test/integration/scheduler/queue_test.go#L287
  - covering the existing behavior of re-queueing with custom resources.

Seems the existing behavior around `MoveAllToActiveOrBackoffQueue` are pretty much covered by these existing tests. 
Regarding the event registration to the event handler, the new test `Test_UnionedGVKs` proves no weird things happen there.
Do you think we need something more?

Also, performance-wise, I checked with scheduler_perf.
(`SchedulingThroughput` is the only metric that we can observe performance degradation of enqueueing if there is.)

master branch:

```
// SchedulingBasic/5000Nodes: the most basic scenario
    {
      "data": {
        "Average": 586.3021871829003,
        "Perc50": 586.3021871829003,
        "Perc90": 586.3021871829003,
        "Perc95": 586.3021871829003,
        "Perc99": 586.3021871829003
      },
      "unit": "pods/s",
      "labels": {
        "Metric": "SchedulingThroughput",
        "Name": "BenchmarkPerfScheduling/SchedulingBasic/5000Nodes/namespace-2"
      }
    },

// PreemptionBasic/500Nodes: It's the scenario that the preemption is involved. (= the requeue must happen during the scenario)
    {
      "data": {
        "Average": 83.13732166013138,
        "Perc50": 7.000241347320932,
        "Perc90": 327.820169984093,
        "Perc95": 327.820169984093,
        "Perc99": 327.820169984093
      },
      "unit": "pods/s",
      "labels": {
        "Metric": "SchedulingThroughput",
        "Name": "BenchmarkPerfScheduling/PreemptionBasic/500Nodes/namespace-2"
      }
    },
```

this branch:

```
    {
      "data": {
        "Average": 590.5119312595991,
        "Perc50": 590.5119312595991,
        "Perc90": 590.5119312595991,
        "Perc95": 590.5119312595991,
        "Perc99": 590.5119312595991
      },
      "unit": "pods/s",
      "labels": {
        "Metric": "SchedulingThroughput",
        "Name": "BenchmarkPerfScheduling/SchedulingBasic/5000Nodes/namespace-2"
      }
    },
//...
    {
      "data": {
        "Average": 83.15325957010394,
        "Perc50": 5.996671949011722,
        "Perc90": 204.4340755303615,
        "Perc95": 204.4340755303615,
        "Perc99": 204.4340755303615
      },
      "unit": "pods/s",
      "labels": {
        "Metric": "SchedulingThroughput",
        "Name": "BenchmarkPerfScheduling/PreemptionBasic/500Nodes2/namespace-2"
      }
    },
```

No performance degradation is observed from them.

**Comment by pohly** [comment]
> Also, performance-wise, I checked with scheduler_perf.

You can do this with `benchstat` now. Run `make test-integration WHAT=./test/integration/scheduler_perf ETCD_LOGLEVEL=warn KUBE_TEST_VMODULE="''" KUBE_TEST_ARGS="-run=^$$ -benchtime=1ns -bench=BenchmarkPerfScheduling/SchedulingBasic -perf-scheduling-label-filter=performance,-fast"`, capture the output, then compare with `benchstat <before>.log <after>.log`.

**Comment by alculquicondor** [comment]
> Do you think we need something more?

I think so. The test `TestCoreResourceEnqueue` seems good, but we should have more scenarios (cases) covering other plugins and a few different events.

benchmarks look good :)

**Review by alculquicondor — COMMENTED** [review]
Ok, I'm happy with this PR as is.

However, the integration tests that I propose in my comment above will be important as we start to move the checks into the plugins hinting functions. But that will be for follow up PRs.

/approve

Leaving LGTM to @pohly

**Comment by pohly** [comment]
This change looks good to me, but before a formal LGTM I'd also like to test it again.

@sanposhiho: can you squash the commits of this PR and update https://github.com/kubernetes/kubernetes/pull/117844 with the latest version of all pending changes?

**Comment by sanposhiho** [comment]
@alculquicondor OK, got it. I'll have a follow-up PR later.
@pohly Sure.

**Review by pohly — APPROVED** [review]
/lgtm

**Review by Huang-Wei — COMMENTED** [review]
Thanks @sanposhiho !

The logic looks good, esp. the logic to honor unschedulablePlugins prior to evaluating schedulingHint, which can guarantee the impl. is performant. Some comments below.

File: pkg/scheduler/framework/runtime/framework.go [hunk 11]
```diff
@@ -545,34 +527,40 @@ func (f *frameworkImpl) expandMultiPointPlugins(logger klog.Logger, profile *con
...
+		// This is to ensure backward compatibility.
+		f.enqueueExtensions = append(f.enqueueExtensions, &defaultEnqueueExtension{pluginName: p.Name()})
 		return
 	}
 
-	events := ext.EventsToRegister()
-	// It's rare that a plugin implements EnqueueExtensions but returns nil.
-	// We treat it as: the plugin is not interested in any event, and hence pod failed by that plugin
-	// cannot be moved by any regular cluster event.
-	if len(events) == 0 {
```
[the remark this answers is not in the document]
Comment (sanposhiho) [reply pkg/scheduler/framework/runtime/framework.go:561]: We keep the equivalent logic in buildQueueingHintMap in scheduler.go and tested here:
https://github.com/sanposhiho/kubernetes/blob/6f8d38406a7f16fc9cc9b72789a9b826105b1b54/pkg/scheduler/scheduler_test.go#L919

(But, for now, regardless of how we treat such plugins, we register all events into EventHandler: https://github.com/kubernetes/kubernetes/pull/118551#discussion_r1225735399)

I'll move the comment to buildQueueingHintMap to state about this case.

File: pkg/scheduler/util/utils.go [hunk 12]
```diff
@@ -159,3 +160,26 @@ func IsScalarResourceName(name v1.ResourceName) bool {
 	return v1helper.IsExtendedResourceName(name) || v1helper.IsHugePageResourceName(name) ||
 		v1helper.IsPrefixedNativeResource(name) || v1helper.IsAttachableVolumeResourceName(name)
 }
+
+// As converts two objects to the given type.
+// Both objects must be of the same type. If not, an error is returned.
+// nil objects are allowed and will be converted to nil.
+func As[T runtime.Object](oldObj, newobj interface{}) (T, T, error) {
```
[the remark this answers is not in the document]
Comment (sanposhiho) [reply pkg/scheduler/util/utils.go:167]: It's gonna be used in every plugin's hint fn like:
https://github.com/sanposhiho/kubernetes/blob/6591db86757ead62e335e8ea73ec93e1b36581b3/pkg/scheduler/framework/plugins/interpodaffinity/plugin.go#L98-L104

(`AsPods()` was reborn as `As[T runtime.Object]()`)

File: pkg/scheduler/internal/queue/scheduling_queue.go [hunk 13]
```diff
@@ -336,6 +353,62 @@ func (p *PriorityQueue) Run(logger klog.Logger) {
 	}, 30*time.Second, p.stop)
 }
 
+// isPodWorthRequeuing calls QueueingHintFn of only plugins registered in pInfo.unschedulablePlugins.
+// If any QueueingHintFn returns QueueImmediately, the scheduling queue is supposed to enqueue this Pod to activeQ.
+// If no QueueingHintFn returns QueueImmediately, but some return QueueAfterBackoff,
+// the scheduling queue is supposed to enqueue this Pod to activeQ/backoffQ depending on the remaining backoff time of the Pod.
+// If all QueueingHintFn returns QueueSkip, the scheduling queue enqueues the Pod back to unschedulable Pod pool
+// because no plugin changes the scheduling result via the event.
+func (p *PriorityQueue) isPodWorthRequeuing(logger klog.Logger, pInfo *framework.QueuedPodInfo, event framework.ClusterEvent, oldObj, newObj interface{}) framework.QueueingHint {
```
[the remark this answers is not in the document]
Comment (sanposhiho) [reply pkg/scheduler/internal/queue/scheduling_queue.go:362]: Leave it to @kerthcet in https://github.com/kubernetes/kubernetes/pull/118929#discussion_r1245061268

File: pkg/scheduler/internal/queue/scheduling_queue.go [hunk 14]
```diff
@@ -747,54 +819,75 @@ func (p *PriorityQueue) AssignedPodUpdated(logger klog.Logger, pod *v1.Pod) {
...
-	p.moveAllToActiveOrBackoffQueue(logger, event, preCheck)
+	p.moveAllToActiveOrBackoffQueue(logger, event, oldObj, newObj, preCheck)
+}
+
+// requeuePodViaQueueingHint tries to requeue Pod to activeQ, backoffQ or unschedulable pod pool based on schedulingHint.
+// It returns the queue name Pod goes.
+//
+// NOTE: this function assumes lock has been acquired in caller
+func (p *PriorityQueue) requeuePodViaQueueingHint(logger klog.Logger, pInfo *framework.QueuedPodInfo, schedulingHint framework.QueueingHint, event string) string {
+	if schedulingHint == framework.QueueSkip {
```
[the remark this answers is not in the document]
Comment (sanposhiho) [reply pkg/scheduler/internal/queue/scheduling_queue.go:847]: Sorry for the confusion, but it can happen after https://github.com/kubernetes/kubernetes/pull/118438.

**Comment by sanposhiho** [comment]
@Huang-Wei Thanks for the additional reviews, I replied some + follow up some in https://github.com/kubernetes/kubernetes/pull/119077 🙏

File: pkg/scheduler/framework/types.go [hunk 15]
```diff
@@ -78,6 +78,62 @@ const (
 	WildCard              GVK = "*"
 )
 
+type ClusterEventWithHint struct {
```
Comment (ahg-g) [inline pkg/scheduler/framework/types.go:81]: The name look odd an too specific; why are we adding a new type? why not place QueueingHintFn inside ClusterEvent?

File: pkg/scheduler/framework/runtime/framework.go [hunk 16]
```diff
@@ -545,34 +527,40 @@ func (f *frameworkImpl) expandMultiPointPlugins(logger klog.Logger, profile *con
...
-			eventToPlugins[evt].Insert(name)
-		}
+// defaultEnqueueExtension is used when a plugin does not implement EnqueueExtensions interface.
+type defaultEnqueueExtension struct {
+	pluginName string
+}
+
+func (p *defaultEnqueueExtension) Name() string { return p.pluginName }
+func (p *defaultEnqueueExtension) EventsToRegister() []framework.ClusterEventWithHint {
+	// need to return all specific cluster events with framework.All action instead of wildcard event
```
Comment (ahg-g) [inline pkg/scheduler/framework/runtime/framework.go:550]: Keeping this in sync with types.go where the types are defined is a source of error; can we define a function in https://github.com/kubernetes/kubernetes/blob/master/pkg/scheduler/framework/types.go that returns the list and we call it from here (e.g., UnrollWildCardResource).

```
events := make([]framework.ClusterEventWithHint)
for res := range framework.UnrollWildCardResource() {
  events = append(events, {Event: framework.ClusterEvent{Resource: res, ActionType: framework.All}})
}
return events
```

File: pkg/scheduler/internal/queue/scheduling_queue.go [hunk 17]
```diff
@@ -747,54 +819,75 @@ func (p *PriorityQueue) AssignedPodUpdated(logger klog.Logger, pod *v1.Pod) {
...
+		p.unschedulablePods.addOrUpdate(pInfo)
+		metrics.SchedulerQueueIncomingPods.WithLabelValues("unschedulable", ScheduleAttemptFailure).Inc()
+		return unschedulablePods
+	}
+
+	pod := pInfo.Pod
+	if p.isPodBackingoff(pInfo) && schedulingHint == framework.QueueAfterBackoff {
```
Comment (ahg-g) [inline pkg/scheduler/internal/queue/scheduling_queue.go:854]: nit, the `schedulingHint == framework.QueueAfterBackoff` condition should be first in the evaluation order

File: pkg/scheduler/internal/queue/scheduling_queue.go [hunk 18]
```diff
@@ -336,6 +353,62 @@ func (p *PriorityQueue) Run(logger klog.Logger) {
...
+			h := hintfn.QueueingHintFn(pod, oldObj, newObj)
+			if h == framework.QueueSkip {
+				continue
+			}
+
+			if h == framework.QueueImmediately {
+				return h
+			}
+
+			// replace queueHint with the returned value,
```
Comment (ahg-g) [inline pkg/scheduler/internal/queue/scheduling_queue.go:401]: we should check that the only option left here is `QueueAfterBackoff`

File: pkg/scheduler/internal/queue/scheduling_queue.go [hunk 19]
```diff
@@ -747,54 +819,75 @@ func (p *PriorityQueue) AssignedPodUpdated(logger klog.Logger, pod *v1.Pod) {
...
+		metrics.SchedulerQueueIncomingPods.WithLabelValues("backoff", event).Inc()
+		return backoffQ
+	}
+
+	if added, _ := p.addToActiveQ(logger, pInfo); added {
+		metrics.SchedulerQueueIncomingPods.WithLabelValues("active", event).Inc()
+		return activeQ
+	}
+
+	p.unschedulablePods.addOrUpdate(pInfo)
```
Comment (ahg-g) [inline pkg/scheduler/internal/queue/scheduling_queue.go:870]: we get here only when there is something completely off, right? if so, please add an error log

**Comment by sanposhiho** [comment]
@ahg-g Thanks for the additional comments, I addressed them in https://github.com/kubernetes/kubernetes/pull/119077

File: pkg/scheduler/framework/types.go (hunk 15 above)
Comment (ahg-g) [reply pkg/scheduler/framework/types.go:81]: I guess the question is: do we ever use ClusterEvent on its own?

Comment (Huang-Wei) [reply pkg/scheduler/framework/types.go:81]: I think in some logic like MoveAllToActiveOrBackoffQueue(), ClusterEvent w/o hintFn is used.

File: pkg/scheduler/internal/queue/scheduling_queue.go [hunk 20]
```diff
@@ -336,6 +353,62 @@ func (p *PriorityQueue) Run(logger klog.Logger) {
...
+			if !pInfo.UnschedulablePlugins.Has(hintfn.PluginName) {
+				continue
+			}
+
+			h := hintfn.QueueingHintFn(pod, oldObj, newObj)
+			if h == framework.QueueSkip {
+				continue
+			}
+
+			if h == framework.QueueImmediately {
```
Comment (kerthcet) [inline pkg/scheduler/internal/queue/scheduling_queue.go:397]: Revisit this logic, if a pod is rejected by several plugins, and only one plugin's queueingHint returns immediately, how can we tell the pod is probably schedulable? It might be rejected by another plugin again.

Then the more reasonable logic here is - If all the unschedulable plugins return __QueueImmediately__, then we'll enqueue the pod for scheduling AFAP.

Comment (sanposhiho) [reply pkg/scheduler/internal/queue/scheduling_queue.go:397]: Let me explain my take on this topic. cc @AxeZhan @ahg-g as we're talking about a similar thing in [another thread](https://github.com/kubernetes/kubernetes/pull/119155#discussion_r1260207009).

What we need is **_a clear definition_** of when to return QueueAfterBackoff and when to return QueueImmediately. Because it's currently unclear ([ref](https://github.com/kubernetes/kubernetes/blob/master/pkg/scheduler/framework/types.go#L109-L121)).
And, we should consider how to handle QueueImmediately based on that.

---

First, BackoffQ is a light way of keeping throughput high by preventing pods that are "permanently unschedulable" from blocking the queue. (quote from https://github.com/kubernetes/kubernetes/pull/117561#discussion_r1189061873)

In my opinion, when to return QueueAfterBackoff and when to return QueueImmediately should be decided not by the possibility of making Pod schedulable or not, but by the reason why the Pod was in the unschedQ now.
We can split the reasons why a Pod puts back into the scheduling queue into these two:
- scheduling failure, like PodAffinity rejects Pod in Filter, NodeResourceFit rejects Pod in Filter, etc.
- non scheduling failure, like DRA needs to wait for the claim to be provisioned, DRA needs to wait for schedulingcontext to be updated by the driver etc.

So, I'm thinking that they equal when to return QueueAfterBackoff (the former) and when to return QueueImmediately (the latter). 

We should always force Pods to honor backoff if they're rejected by the scheduling failure. 
Because, in any cases, we cannot say that this Pod will 100% get scheduled in the next scheduling cycle because the cluster's situation keeps changing from moment to moment and any plugins can reject Pod in the next scheduling cycle. 

OTOH, it's OK to ignore backoff if they're not rejected by the scheduling failure. Such Pods don't have the obligation to go through backoffQ because they're not rejected by the scheduling failure.

These are my thought. And specifically speaking, DRA is the only plugin among in-tree plugins that causes Pods to be pushed back into the scheduling queue by non scheduling failure. Meaning, DRA is the only plugin that can return QueueImmediately. Other plugins like NodeAffinity we discussed are all Filter plugins, and cause Pods to be pushed back into the scheduling queue by scheduling failure. Meaning they all can return QueueAfterBackoff only.

What do you all think?

Comment (sanposhiho) [reply pkg/scheduler/internal/queue/scheduling_queue.go:397]: > if DRA was NOT the only reason why a pod is unschedulable, then does it make sense to queue immediately? 

It makes sense if we follow my guidance above.
Regardless of whether DRA is the only reason or there are many other unschedulable plugins registered for a Pod as well, either way, we cannot be completely certain that a Pod will be successfully scheduled when an arrived event resolves DRA's past rejection. That depends on how each plugin is implemented and how the cluster situation gets changed, which we cannot completely manage.

So, the only thing we do is to determine whether the past rejection is due to scheduling failure or non scheduling failure, and then decide to honor backoff or skip backoff only based on that. We can skip backoff when non scheduling failure, because there is no reason that such Pods experience backoff, not because it's a higher possibility that a Pod can be schedulable in the next scheduling cycle.

> Do we have the ability to distinguish that? I suppose not.

Currently No. But, given QueueImmediately only comes from the DRA plugin among in-tree plugins at least for now, we can rethink the improvement when needed based on the usecase at that time.

Comment (kerthcet) [reply pkg/scheduler/internal/queue/scheduling_queue.go:397]: But the `DRA needs to wait for the claim to be provisioned` is also a scheduling failure and can happen together with other failure plugins. We have similar situation in volume binding when the PV isn't exist yet.

Another input: If we return QueueSkip in one failed plugin, can we tell that the pod doesn't worth requeueing, the hint should not be override by QueueAfterBackoff (I didn't consider `QueueImmediately` here).

Comment (sanposhiho) [reply pkg/scheduler/internal/queue/scheduling_queue.go:397]: To be accurate, let me rephrase it that the Scheduler needs to wait for claim to be allocated/reserved. (instead of created) DRA is different from PV as you can see it has Reserve() to do that. In DRA, there is a concept WaitForFirstConsumer.
https://github.com/kubernetes/kubernetes/blob/v1.27.1/pkg/scheduler/framework/plugins/dynamicresources/dynamicresources.go#L700

That's obviously not a scheduling failure, that just waiting for the external component to do something.

> Another input: If we return QueueSkip in one failed plugin, can we tell that the pod doesn't worth requeueing, the hint should not be override by QueueAfterBackoff

No, we shouldn’t do that. We need to remember that the unschedulable plugins (set by the filter plugins) mean that all of them rejected **_some_** Nodes, doesn’t mean that all of them rejected all Nodes. So even if pluginA and pluginB are registered as unschedulable plugins, maybe we only need to solve the failure of either of them to get Pod scheduled.

Comment (sanposhiho) [reply pkg/scheduler/internal/queue/scheduling_queue.go:397]: I have one proposal from the discussion here: https://github.com/kubernetes/kubernetes/pull/119517
PTAL when get a chance. 🙏
